flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java

```diff
@@ -365,4 +409,88 @@ public String toString() {
                 "%s: current checkpoint: %d, current barriers: %d, open channels: %d",
                 taskName, currentCheckpointId, numBarriersReceived, numOpenChannels);
     }
+
+    class ContextImpl implements BarrierHandleState.Context {
+        private final Map<InputChannelInfo, Integer> sequenceNumberInAnnouncedChannels =
+                new HashMap<>();
+
+        private final Set<InputChannelInfo> blockedChannels = new HashSet<>();
```

Comment:
Maybe those state fields should be moved to `BarrierHandleState`, while `ContextImpl` would become a stateless class - just a proxy for action calls?